cl_render.lua

```diff
@@ -0,0 +1,76 @@
+-- optimizations
+local vec = vec
+local Wait = Wait
+local format = string.format
+local RemoveBlip = RemoveBlip
+local PlayerPedId = PlayerPedId
+local IsHudHidden = IsHudHidden
+local SetTextFont = SetTextFont
+local CreateThread = CreateThread
+local SetTextScale = SetTextScale
+local SetTextOutline = SetTextOutline
+local GetEntityCoords = GetEntityCoords
+local EndTextCommandDisplayText = EndTextCommandDisplayText
+local BeginTextCommandDisplayText = BeginTextCommandDisplayText
+local AddTextComponentSubstringPlayerName = AddTextComponentSubstringPlayerName
+-- end optimizations
+
+local nearestPostalText = ""
+
+-- recalculate current postal
+CreateThread(function()
+    local delay = config.updateDelay
+    if not delay or tonumber(delay) <= 0 then
+        error("Invalid render delay provided, it must be a number > 0")
+    end
+
+    local postals = postals
+    local deleteDist = config.blip.distToDelete
+    local formatTemplate = config.text.format
+    local _total = #postals
+
+    while true do
+        local coords = GetEntityCoords(PlayerPedId())
+        local _nearestIndex, _nearestD
+        coords = vec(coords[1], coords[2])
+
+        for i = 1, _total do
+            local D = #(coords - postals[i][1])
```

Comment:
This internally performs a `sqrt`, so it's probably better to switch over to checking distance magnitudes in these loops. You can either use the built in function `Vdist2` or writing something like:
```lua
local v = coords - postals[i][1]
local D = v.x^2 + v.y^2 + v.z^2
```

then after the loop you can find the `sqrt` for finding canonical distance from the postal

Reply:
I'm not sure I understand what you mean by this, could you clarify why `sqrt` is necessary at all?

The `#(vecA - vecB)` syntax does an optimized distance calculation, and should return the distance in game units (meters). `Vdist2` and custom math will likely be slower than native Lua implemented in C.

Reply:
The point that I'm trying to make is that a `sqrt` wouldn't be necessary for comparisons between distances, only for the final display of distance. It's explained pretty well on the [C# Vector3 Docs](https://github.com/citizenfx/fivem/blob/master/code/client/clrcore/Math/Vector3.cs#L649) why you would want to use distance magnitude over distance.

Do some testing, and see if using `#(vec1 - vec2)` is faster than writing it out in lua or using `Vdist2`. Or maybe it doesn't make any difference at all :shrug:

Reply:
I'll do some more in-depth performance testing and get back to you!

Reply:
Here are the results I got from my performance testing. These times do include the network time (as I used events), but it should be negligible for localhost. (The repo of my perf testing code is [available here](https://github.com/GoatG33k/FiveM-dist-calc-perf-testing))

| **Code**                                                 | **Iterations** | **Mean Time** | **Mean ops/sec**  |
|----------------------------------------------------------|----------------|---------------|-------------------|
| `local D = #(vecA - vecB)`                               | 5x @ 10mil     | 447.6ms       | ~22,388.4 ops/sec |
| `local v = vecA - vecB; local D = v.x^2 + v.y^2 + v.z^2` | 5x @ 10mil     | 726.0ms       | ~13,783.8 ops/sec |
| `local D = Vdist(a[1], a[2], a[3], b[1], b[2], b[3])`    | 5x @ 10mil     | 1682.0ms      | ~5,950.4 ops/sec  |
| `local D = Vdist2(a[1], a[2], a[3], b[1], b[2], b[3])`   | 5x @ 10mil     | 1635.0ms      | ~6,139.4 ops/sec  |